selfservice/sessiontokenexchange/handler.go

```diff
@@ -0,0 +1,143 @@
+// Copyright © 2023 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package sessiontokenexchange
+
+import (
+	"net/http"
+
+	"github.com/gofrs/uuid"
+	"github.com/julienschmidt/httprouter"
+
+	"github.com/ory/herodot"
+	"github.com/ory/kratos/driver/config"
+	"github.com/ory/kratos/session"
+	"github.com/ory/kratos/x"
+)
+
+const (
+	RouteExchangeCodeForSessionToken = "/self-service/exchange-code-for-session-token" // #nosec G101
+)
+
+type (
+	handlerDependencies interface {
+		PersistenceProvider
+		config.Provider
+		x.WriterProvider
+		session.PersistenceProvider
+	}
+
+	HandlerProvider interface {
+		SessionTokenExchangeHandler() *Handler
+	}
+	Handler struct {
+		d handlerDependencies
+	}
+)
+
+func NewHandler(d handlerDependencies) *Handler {
+	return &Handler{d: d}
+}
+
+func (h *Handler) RegisterPublicRoutes(public *x.RouterPublic) {
+	public.GET(RouteExchangeCodeForSessionToken, h.exchangeCode)
+}
+
+func (h *Handler) RegisterAdminRoutes(admin *x.RouterAdmin) {
+	admin.GET(RouteExchangeCodeForSessionToken, x.RedirectToPublicRoute(h.d))
+}
+
+// Exchange Session Token Parameters
+//
+// swagger:parameters exchangeSessionToken
+//
+//nolint:deadcode,unused
+//lint:ignore U1000 Used to generate Swagger and OpenAPI definitions
+type exchangeSessionToken struct {
+	// The Login Flow ID
+	//
+	// The value for this parameter comes from `flow` URL Query parameter sent to your
+	// application (e.g. `/login?flow=abcde`).
+	//
+	// required: true
+	// in: query
+	ID string `json:"id"`
+
+	// The Session Token Exchange Code
+	//
+	// required: true
+	// in: query
+	SessionTokenExchangeCode string `json:"code"`
+}
+
+// The Response for Registration Flows via API
+//
+// swagger:model successfulCodeExchangeResponse
+type codeExchangeResponse struct {
+	// The Session Token
+	//
+	// A session token is equivalent to a session cookie, but it can be sent in the HTTP Authorization
+	// Header:
+	//
+	// 		Authorization: bearer ${session-token}
+	//
+	// The session token is only issued for API flows, not for Browser flows!
+	Token string `json:"session_token,omitempty"`
+
+	// The Session
+	//
+	// The session contains information about the user, the session device, and so on.
+	// This is only available for API flows, not for Browser flows!
+	//
+	// required: true
+	Session *session.Session `json:"session"`
+}
+
+// swagger:route GET /self-service/login/exchange-session-token frontend exchangeSessionToken
```

Comment:
How about

```suggestion
// swagger:route POST /self-service/login/exchange frontend exchangeLoginCodeForSessionToken
```

Reply:
I renamed this to `/self-service/exchange-code-for-session-token` because it is now flow-agnostic. WDYT?

Reply:
👍

Reply:
Or, in that case, it might make more sense to have this under `/sessions/exchange`?

Reply:
naming-wise the `/self-service/exchange-code-for-session-token` API doesn't really match any other endpoint exposed by ory - lots of words and hyphens. Maybe `/sessions/token-exchange`?

Reply:
fixed